### PR TITLE
Improve rotate then insert selected bits on Z

### LINKED ISSUE #3212: Improve rotate then insert selected bits on Z
While working on a miscellaneous item in OpenJ9 I had to use an `RISBG` instruction to perform an and and a shift. Since the `RISBG` instructions are slightly confusing in PoPs I wrote a quick Java program and compiled it to validate that I used the correct instruction. Here is the Java program I used and how I invoked it:

```
> cat Test.java
class Test
{
    public static void main(String[] args)
    {
        foo(5L);
        foo(5L);
    }

    public static long foo(long x)
    {
        return ((long)(x & 0xF0) >> 4L);
    }
}

>./bin/java -Xjit:'{Test.foo*}(count=0,tracefull,log=log.trace)' Test
```

It generated the following trees:

```
------------------------------
 n8n      (  0)  lreturn                                                                              [     0x3ff96804ce0] bci=[-1,7,11] rc=0 vc=130 vn=- li=2 udi=- nc=1
 n7n      (  0)    lshr (in GPR_0018) (highWordZero X>=0 cannotOverflow )                             [     0x3ff96804c90] bci=[-1,6,11] rc=0 vc=130 vn=- li=2 udi=- nc=2 flg=0x5100
 n5n      (  0)      land (in GPR_0016) (highWordZero X>=0 cannotOverflow )                           [     0x3ff96804bf0] bci=[-1,4,11] rc=0 vc=130 vn=- li=2 udi=- nc=2 flg=0x5100
 n10n     (  0)        ==>lRegLoad (in GPR_0016) (cannotOverflow SeenRealReference )
 n4n      (  0)        lconst 240 (highWordZero X!=0 X>=0 )                                           [     0x3ff96804ba0] bci=[-1,1,11] rc=0 vc=130 vn=- li=2 udi=- nc=0 flg=0x4104
 n6n      (  0)      iconst 4 (X!=0 X>=0 )                                                            [     0x3ff96804c40] bci=[-1,5,11] rc=0 vc=130 vn=- li=2 udi=- nc=0 flg=0x104
------------------------------

 [     0x3ff968930e0]   RISBGN  GPR_0016,GPR_0016,56,187,0
 [     0x3ff968af000]   SRAG    GPR_0018,GPR_0016,4
 [     0x3ff968af8e0]   ASSOCREGS
 [     0x3ff968af190]   RET
 POST:
 {GPR2:GPR_0018:R}
```

This sequence can be improved. I think the `SRAG` instruction is redundant as we can shift right and zero out everything else just with the `RISBG` instruction. This sequence should be improved to eliminate the `SRAG` instruction.

### PR DESCRIPTION
**Pull request by AidanHa**
Add a case in genericLongShiftSingle that would generate
a RISBG command for a LSHR->LAND pattern

Add a similar case for LSHL->LAND pattern

Add a similar case for the unsigned variants of the
instruction(LUSHR and LUSHL)

Rename tryToReplaceLongAndWithRotateInstruction() to
tryToReplaceShiftLandWithRotateInstruction()

Allocate target register properly in
tryToReplaceShiftLandWithRotateInstruction()

Add respective Tril test for these changes

Fixes #3212

Signed-off-by: Aidan Ha <qbha@edu.uwaterloo.ca>

### CODE REVIEW

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 1]
```diff
@@ -1223,6 +1223,30 @@ genericLongShiftSingle(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::
...
             }
+         else if (node->getOpCodeValue() == TR::lshr)
+            {
+            // Generate RISBGN for lshr + land sequence
+            if (firstChild->getOpCodeValue() == TR::land)
+               {
+               if (value < 64)
+                  {
+                  //If the rotated value contains 8 bits or less (So it fits in the instruction)
+                  srcReg = cg->evaluate(firstChild->getFirstChild());
```
[the remark this answers is not in the document]
Comment (AidanHa) [reply compiler/z/codegen/BinaryEvaluator.cpp]: I ended up not evaluating `firstChild->getFirstChild()`, since it will be evaluated in  `tryToReplaceLongAndWithRotateInstruction`

**Comment by AidanHa** [comment]
There's currently a set of test cases that I can't seem to find a solution around:

image


No matter what I tried doing, I couldn't pass this case due to the fact that we are using a rotate instruction and we have no information of `0x000...1` (which is a runtime value). Anyone have a solution that I might be missing?

**Comment by fjeremic** [comment]
@AidanHa the tree you specified can be implemented with `RISBG`. Let's use this example:

```
lshr
  land
    x
    lconst 0xFF..FF
  iconst 1
```

This can get reduced down to the following instruction:

```
RISBG x,x,0,190,63
```

Deciphering this what it says is to select bits 0 - 62 inclusive of `x` then rotate those bits by 63 to the left, which is the equivalent of rotating them 1 to the right. If you replace `x` with 0x00..01 as in your example the final output should be 0 after executing that instruction.

**Comment by AidanHa** [comment]
Up for another round of reviews!

File: compiler/z/codegen/OMRTreeEvaluator.hpp [hunk 2]
```diff
@@ -235,13 +235,21 @@ class OMR_EXTENSIBLE TreeEvaluator: public OMR::TreeEvaluator
     *
     *   \param cg
     *      The code generator used to generate the instructions
+    * 
+    *   \param shiftAmount
+    *      If using function to replace a shift+and combination, this is the variable that holds how much we want to shift.
+    *      If we are using this function to replace a standalone 'and', then this variable should be zero.
```
Comment (fjeremic) [inline compiler/z/codegen/OMRTreeEvaluator.hpp]: The documentation for this parameter is incomplete. The parameter type is signed on purpose but it does not tell the user of the API what it means for a value to be positive vs. negative. How about:

```
    *   \param shiftAmount
    *      The number of bits we should shift the result of the logical AND. A positive value represents a left shift,
    *      a negative value represents a right shift, and a value of zero represents no shift.
```

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 3]
```diff
@@ -1898,7 +1949,52 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
+
+               if (shiftAmount <= 0)
+                  {
+                  if (msBit - shiftAmount > 63)
+                     {
+                     return NULL;
+                     }
+                  else
+                     {
+                     rangeStart = msBit - shiftAmount;
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: This looks incorrect. If I understand a `shiftAmount < 0` implies a right shift so we are modifying the selected range by the shift amount? What happens for the following tree:

```
lushr
  land
    x
    lconst 0x4000000000000000
  lconst 1
```

I think `msBit == lsBit == 1` in this case, so `rangeStart = 1 - (-1) == 2` so we'll end up selecting the third most significant bit of 0x4000000000000000 which is 0 and we should get the wrong result here.

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 4]
```diff
@@ -1898,7 +1949,52 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
                // The bits in that range are preserved as is
-               generateRIEInstruction(cg, opCode, node, targetReg, sourceReg, msBit, 0x80 + lsBit, 0);
+               int32_t rangeStart = msBit;
+               int32_t rangeEnd = lsBit;
+
+               if (shiftAmount <= 0)
+                  {
+                  if (msBit - shiftAmount > 63)
+                     {
+                     return NULL;
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: These checks should happen before we set `doTransformation = true` because we've already printed in the log that we are about to perform the transformation.

File: compiler/z/codegen/BinaryEvaluator.cpp (hunk 3 above)
Comment (fjeremic) [reply compiler/z/codegen/BinaryEvaluator.cpp]: I don't think your changes should be modifying selection bits, i.e. `rangeStart` and `rangeEnd` should not exist. All your modifications should do is add the rotate amount to the existing `generateRIEInstruction` instead of passing 0, and logic should be added to ensure that the rotate is really a shift.

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 5]
```diff
@@ -1205,13 +1205,15 @@ genericLongShiftSingle(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::
       if (TR::Compiler->target.cpu.getSupportsArch(TR::CPU::z10))
          {
          // Generate RISBG for lshl + i2l sequence
-         if (node->getOpCodeValue() == TR::lshl)
+         if (node->getOpCodeValue() == TR::lshl || node->getOpCodeValue() == TR::lushl)
             {
-            if (firstChild->getOpCodeValue() == TR::i2l && firstChild->isSingleRefUnevaluated() && (firstChild->isNonNegative() || firstChild->getFirstChild()->isNonNegative()))
+            bool isSignedLeftShift = (node->getOpCodeValue() == TR::lshl) ? 1 : 0;
+            if (isSignedLeftShift == 0 && (firstChild->getOpCodeValue() == TR::i2l && firstChild->isSingleRefUnevaluated() && (firstChild->isNonNegative() || firstChild->getFirstChild()->isNonNegative())))
```
[the remark this answers is not in the document]
Comment (AidanHa) [reply compiler/z/codegen/BinaryEvaluator.cpp]: Right, so I'll simply remove the restriction I added.

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 6]
```diff
@@ -1222,10 +1224,50 @@ genericLongShiftSingle(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::
...
+               if (value < 64)
+                  {
+                  //Rotated value must be 6 bits or less in length in order to fit into the instruction.
+                  if (trgReg = TR::TreeEvaluator::tryToReplaceShiftLandWithRotateInstruction(firstChild, cg, (int8_t)value, isSignedLeftShift))
+                     {
+                     node->setRegister(trgReg);
+                     cg->decReferenceCount(firstChild->getFirstChild());
+                     cg->decReferenceCount(firstChild->getSecondChild());
+                     cg->decReferenceCount(firstChild);
+                     cg->decReferenceCount(secondChild);
```
[the remark this answers is not in the document]
Comment (AidanHa) [reply compiler/z/codegen/BinaryEvaluator.cpp]: Makes sense. How come there was no decReferenceCount before in the older `tryToReplaceLongAndWithRotateInstruction()`?

[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/BinaryEvaluator.cpp]: > How come there was no decReferenceCount before in the older tryToReplaceLongAndWithRotateInstruction()?

Likely something that was missed during code reviews when the original code was merged.

[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/BinaryEvaluator.cpp]: Yeah, looks like there was only one use of this routine in the past, and the caller made sure to decrement the ref count. So we didn't see a bug : https://github.com/eclipse/omr/blob/4dea2b3869e9e16b1ad3e007ec508bc38bc947c2/compiler/z/codegen/BinaryEvaluator.cpp#L3349-L3357

Not great, especially now that there are multiple callers.

File: compiler/z/codegen/BinaryEvaluator.cpp (hunk 3 above)
Comment (AidanHa) [reply compiler/z/codegen/BinaryEvaluator.cpp]: >so we'll end up selecting the third most significant bit of 0x4000000000000000 which is 0.

I think we're actually selecting the third most significant bit of x instead... I have a test case for this specific value, and it does not fail... Unless there's something I'm not getting here?

Comment (AidanHa) [reply compiler/z/codegen/BinaryEvaluator.cpp]: > I don't think your changes should be modifying selection bits, i.e. rangeStart and rangeEnd should not exist. All your modifications should do is add the rotate amount to the existing generateRIEInstruction instead of passing 0, and logic should be added to ensure that the rotate is really a shift.

`rangeStart` and `rangeEnd` merely exist for the "overflow issue (I added some comments on this in my next commit), where the the optimization will still work even if some of the 1's between the msb and lsb are shifted out.

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 7]
```diff
@@ -1222,10 +1224,40 @@ genericLongShiftSingle(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::
...
+                  }
+               }
+            }
+         else if (node->getOpCodeValue() == TR::lshr || node->getOpCodeValue() == TR::lushr)
+            {
+            // Generate RISBGN for (lshr + land) and (lushr + land) sequences
+            const bool isSignedRightShift = node->getOpCodeValue() == TR::lshl;
+            if (firstChild->getOpCodeValue() == TR::land)
+               {
+               if (trgReg = TR::TreeEvaluator::tryToReplaceShiftLandWithRotateInstruction(firstChild, cg, -(value), isSignedRightShift))
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: Unnecessary parenthesis around `value` in the arguments.

File: compiler/z/codegen/BinaryEvaluator.cpp (hunk 4 above)
Comment (fjeremic) [reply compiler/z/codegen/BinaryEvaluator.cpp]: Now that we've moved these checks above these ones can be removed.

File: compiler/z/codegen/BinaryEvaluator.cpp (hunk 3 above)
Comment (fjeremic) [reply compiler/z/codegen/BinaryEvaluator.cpp]: > I think we're actually selecting the third most significant bit of x instead

That is exactly the problem. We should be selecting the second most significant bit of (unsigned) _x_ and then shifting right by 1 bit, not selecting the third bit and shifting right by 1. If you replace _x_ with 0x4000000000000000 then the result of that tree when evaluated should be 0x2000000000000000. However I believe your current code will output 0.

Can you generate a log for this case and paste the evaluation of such a tree? I'd like to see the exact `RISBG` instruction that gets generated. I'd also be curious to see the execution, so I'll ask that you hop into gdb and paste the register value before and after the `RISBG` instruction.

Comment (fjeremic) [reply compiler/z/codegen/BinaryEvaluator.cpp]: Ok, apologies I got things mixed up. Just to confirm the rotation happens first, then the bits are selected?

So in the case above we would would rotate _x_ by 63 to the left, meaning a shift right, then we would select the third most significant bit of the _rotated x_ and insert it into the result while zeroing out all the other bits?

Can you generate a log with the tree when _x_ = 0x4000000000000000 and paste the register values before and after the `RISBG` instruction?

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 8]
```diff
@@ -1871,35 +1930,85 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
             //    will require two NI** instructions. Hence we are better off using a single RISBG
+            //
+            // "shiftAmount == 0"
+            //    For cases where we have zeros surrounded by one, if the shift value is not zero
+            //    then the optimization will not work.
 
-            if (firstChild->getReferenceCount()> 1 || (lOnes < 32 && tOnes < 32))
+            if ((firstChild->getReferenceCount()> 1 || (lOnes < 32 && tOnes < 32)) && shiftAmount == 0)
                {
+                  // Similar to above, if the shift amount is NOT zero, then we must account for the shifted range
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: Indentation problems here need to be corrected (for the entire block).

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 9]
```diff
@@ -1871,35 +1930,85 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
-
                // if possible then use the instruction that doesn't set the CC as it's faster
                TR::InstOpCode::Mnemonic opCode = TR::Compiler->target.cpu.getSupportsArch(TR::CPU::zEC12) ? TR::InstOpCode::RISBGN : TR::InstOpCode::RISBG;
 
-               // this instruction sets the rotation factor to 0 and sets the zero bit(0x80).
-               // So it's effectively zeroing out every bit except the inclusive range of lsBit to msBit
-               // The bits in that range are preserved as is
-               generateRIEInstruction(cg, opCode, node, targetReg, sourceReg, msBit, 0x80 + lsBit, 0);
-
+               // If the shift amount is zero, this instruction sets the rotation factor to 0 and sets the zero bit(0x80).
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: Indentation problems here need to be corrected (for the entire block).

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 10]
```diff
@@ -1848,10 +1888,29 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
+                     return NULL;
+                     }
+                  }
+               else if (shiftAmount > 0)
+                  {
+                  if (lsBit - shiftAmount < 0)
+                     {
+                     return NULL;
+                     }
+                  }
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: How about:

```
if ((shiftAmount < 0 && msBit - shiftAmount > 63) ||
    (shiftAmount > 0 && lsBit - shiftAmount < 0))
   {
   return NULL;
   }
```

instead of these 14 lines. Also the comment needs to be formatted. There seems to be an extra 3 spaces.

File: compiler/z/codegen/BinaryEvaluator.cpp [hunk 11]
```diff
@@ -1871,35 +1930,85 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
...
```
Comment (fjeremic) [inline compiler/z/codegen/BinaryEvaluator.cpp]: Similar as above.

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp [hunk 12]
```diff
@@ -1188,7 +1215,9 @@ TEST_P(UInt64MaskThenShift, UsingLoadParam) {
 
 INSTANTIATE_TEST_CASE_P(ShiftAndRotateTest, UInt64MaskThenShift, ::testing::Combine(
     ::testing::ValuesIn(static_cast< std::vector<std::tuple<std::tuple<uint64_t, uint64_t>, int32_t>> (*) (void)>(test_mask_then_shift_values)()),
-    ::testing::Values(std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushr", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_right))
+    ::testing::Values(
+        std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushr", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_right)),
+        std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushl", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_left))
```
Comment (fjeremic) [inline fvtest/compilertriltest/ShiftAndRotateTest.cpp]: It seems the tests here are now a superset of the tests in LongAndAsRotateTest. Can we now deprecate that test bucket so we don't have duplication?

[1] https://github.com/eclipse/omr/blob/master/fvtest/compilertriltest/LongAndAsRotateTest.cpp

Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: It seems as if all of the tests in `LongAndAsRotateTest` are a subset of the tests added here. Should I just be removing the file entirely?

Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: If `LongAndAsRotateTest` is a subset of `ShiftAndRotateTest` all that needs to be done is to remove the file.

**Comment by fjeremic** [comment]
@genie-omr build all

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 12 above)
Comment (aviansie-ben) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: I'm not sure I agree that `LongAndAsRotateTest` is now a subset of `ShiftAndRotateTest`... It looks to me like `LongAndAsRotateTest` contains test cases that test mixes of conversions and masks, while the new test cases added in `ShiftAndRotateTest` are testing mixes of masks and shifts. While these do trigger similar optimizations, I don't really see any duplication between the two...

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp [hunk 13]
```diff
@@ -1017,14 +1017,38 @@ std::vector<T> test_masks();
 
 static uint64_t mask64Values[] =
     {
-    0x0000000000000000,
-    0xffffffffffffffff,
+    0x0,
```
Comment (aviansie-ben) [inline fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Is there a reason to be adding this many new 64-bit mask values? Are these testing corner cases that weren't covered by the previous values? The mask-then-shift tests are using the cross-product of three different sets of test values (one generic value, one mask, and one shift), so it's really easy to get combinatorial explosion by adding too many here...

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp [hunk 14]
```diff
@@ -1188,7 +1226,9 @@ TEST_P(UInt64MaskThenShift, UsingLoadParam) {
 
 INSTANTIATE_TEST_CASE_P(ShiftAndRotateTest, UInt64MaskThenShift, ::testing::Combine(
     ::testing::ValuesIn(static_cast< std::vector<std::tuple<std::tuple<uint64_t, uint64_t>, int32_t>> (*) (void)>(test_mask_then_shift_values)()),
-    ::testing::Values(std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushr", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_right))
+    ::testing::Values(
+        std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushr", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_right)),
+        std::make_tuple<const char*, uint64_t (*) (uint64_t, uint64_t, int32_t)>("lushl", static_cast<uint64_t (*) (uint64_t, uint64_t, int32_t)>(mask_then_shift_left))
```
Comment (aviansie-ben) [inline fvtest/compilertriltest/ShiftAndRotateTest.cpp]: I don't really think we should be testing `lushl`. That opcode is being deprecated since it's functionally equivalent to its signed counterpart (see #2657).

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 12 above)
Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: @AidanHa can you comment on the above?

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 13 above)
Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: These were all of the corner cases that were accounted for in the removal of the `LongAndAsRotate` test. I guess that these corner cases were only technically tested previously with a shift value of zero.

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 14 above)
Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Right, I only added the lushl test because I've added support for a `lushl`+`land` combination. @fjeremic Should we still be testing for `lushl`?

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 12 above)
Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: I think the new tests are testing the `LongAndAsRotate` tests with a shift of `zero`. However, it does not test for conversions, so maybe keeping the `LongAndAsRotate` tests for test for conversions might be the better choice.

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 14 above)
Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Left shifts have no bearing on the sign bit, it's the signedness of the right shift we have to be careful about. Yes we can remove the unsigned left shift tests.

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 13 above)
Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: The new mask values come from [1] for which there is a comment explaining all the different cases we are trying to cover. If the newly added masks do not cover additional cases, then yes I agree we don't need to add them. @AidanHa can you go through the existing masks before your change and identify where we are adding coverage with all the new mask values? Was there a case the original mask values were not capturing?

Perhaps there is redundancy to some level. I'm not sure.

[1] https://github.com/eclipse/omr/blob/d309a6eaafd43ca1040175a09c0a68f30fe4357e/fvtest/compilertriltest/LongAndAsRotateTest.cpp#L32-L61

Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Most of the newly added mask values came from removing the `LongAndAsRotate` test. Now that We've decided to leave that test alone for now, we can remove a lot of these values. I'll have a push soon that will remove all of the redundant values.

Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: The ones that will be added were edge cases captured while testing this feature. I've also added some cases where the new feature should NOT use the optimization.

Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: @aviansie-ben I think the number of new cases has been reduced. Could you please re-review? Let me know which threads can be resolved.

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp [hunk 15]
```diff
@@ -1017,14 +1017,27 @@ std::vector<T> test_masks();
 
 static uint64_t mask64Values[] =
     {
-    0x0000000000000000,
-    0xffffffffffffffff,
+    0x0,
+    0x1,
+    0x2,
+    0x3E,
```
Comment (aviansie-ben) [inline fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Just a minor nit: It would probably be best to format these values to be consistent with the other values in this list (lowercase hex digits with left-padding zeroes).

Comment (AidanHa) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Fixed!

**Comment by fjeremic** [comment]
@genie-omr build all

File: fvtest/compilertriltest/ShiftAndRotateTest.cpp (hunk 12 above)
Comment (fjeremic) [reply fvtest/compilertriltest/ShiftAndRotateTest.cpp]: Opened up #4473 to track this.
